### PR TITLE
version: creating ncs_version.h with input from VERSION.

### PR DESCRIPTION
**Pull request by tejlmand**
This commit generates a ncs_version.h which can be included in source
code.
version.h is generated from version.h.in with data from VERSION.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

**Comment by tejlmand** [comment]
Produces:
```
$  cat (find build/zephyr -name ncs_version.h)
#ifndef _NCS_VERSION_H_
#define _NCS_VERSION_H_

#define NCS_VERSION_CODE 66403
#define NCS_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define NCS_VERSION_NUMBER  0x10363
#define NCS_VERSION_MAJOR   1
#define NCS_VERSION_MINOR   3
#define NCS_PATCHLEVEL      99
#define NCS_VERSION_STRING  "1.3.99"

#endif /* _NCS_VERSION_H_ */
```

**Comment by tejlmand** [comment]
and for an `rc1`:
```
$ cat (find build/zephyr -name ncs_version.h)
#ifndef _NCS_VERSION_H_
#define _NCS_VERSION_H_

#define NCS_VERSION_CODE 66403
#define NCS_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define NCS_VERSION_NUMBER  0x10363
#define NCS_VERSION_MAJOR   1
#define NCS_VERSION_MINOR   3
#define NCS_PATCHLEVEL      99
#define NCS_VERSION_STRING  "1.3.99-rc1"

#endif /* _NCS_VERSION_H_ */
```

**Review by carlescufi — COMMENTED** [review]
Looks good, but could we get `conf.py` to also pick it up from here? either as an environment variable or something similar?

something like this: 
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/conf.py#L83

**Comment by tejlmand** [comment]
> Looks good, but could we get `conf.py` to also pick it up from here? either as an environment variable or something similar?
 
Done.

**Comment by tejlmand** [comment]
@ru-fu we now extract the version and release info for the docs from VERSION:
https://github.com/nrfconnect/sdk-nrf/blob/master/VERSION

**Comment by ru-fu** [comment]
> 
> 
> @ru-fu we now extract the version and release info for the docs from VERSION:
> https://github.com/nrfconnect/sdk-nrf/blob/master/VERSION

There are several other places where we need to update the version for a release, but it's a good start. :)

**Review by ru-fu — APPROVED** [review]

**Comment by tejlmand** [comment]
> > @ru-fu we now extract the version and release info for the docs from VERSION:
> > https://github.com/nrfconnect/sdk-nrf/blob/master/VERSION
> 
> There are several other places where we need to update the version for a release, but it's a good start. :)

@ru-fu Bring them on :smile:

**Comment by ru-fu** [comment]
> 
> 
> > > @ru-fu we now extract the version and release info for the docs from VERSION:
> > > https://github.com/nrfconnect/sdk-nrf/blob/master/VERSION
> > 
> > 
> > There are several other places where we need to update the version for a release, but it's a good start. :)
> 
> @ru-fu Bring them on 😄

https://github.com/nrfconnect/sdk-nrf/pull/3073/files
Plus "Update documentation version for the different doc sets (in the doc/*/conf.py files)" - we also have versions for mcuboot and nrfxlib, but those weren't updated for 1.3.2.

File: doc/nrf/conf.py [hunk 1]
```diff
@@ -112,9 +113,18 @@
...
-version = '1.3.99'
+version = 'unknown'
+
 # The full version, including alpha/beta/rc tags.
-release = '1.3.99'
+release = 'unknown'
+
+with open(os.path.join(NRF_BASE, 'VERSION')) as version_file:
+    for line in version_file:
+        match = re.search(r'^(\d*\.\d*\.\d*)(-[^ \r\n\t]*)?', line)
```
Comment (utzig) [inline doc/nrf/conf.py:123]: I believe the second match could be `(-\S*)?`

**Review by utzig — APPROVED** [review]
LGTM. I didn't know cmake could do math and bitshifts!

File: doc/nrf/conf.py (hunk 1 above)
Comment (tejlmand) [reply doc/nrf/conf.py]: Thanks for the suggestion.
Done.

**Comment by tejlmand** [comment]
> > > > @ru-fu we now extract the version and release info for the docs from VERSION:
> > > > https://github.com/nrfconnect/sdk-nrf/blob/master/VERSION
> > > 
> > > 
> > > There are several other places where we need to update the version for a release, but it's a good start. :)
> > 
> > 
> > @ru-fu Bring them on smile
> 
> https://github.com/nrfconnect/sdk-nrf/pull/3073/files

Thanks for the link.
One question, why is this needed:
https://github.com/nrfconnect/sdk-nrf/blob/343b938e0ae7e3ad4b77b05be9481901fae51d3e/doc/nrf/shortcuts.txt#L3
when we have `conf.py` ?

it seems most places are updated to use `|release|` or similar, and those that aren't, like [versions.json](https://github.com/nrfconnect/sdk-nrf/blob/master/doc/versions.json) and release-notes-x.y.z must have fixed numbers.

One observation I did is this line:
```
for development in |release| for prototyping
```
in [README.rst](https://github.com/nrfconnect/sdk-nrf/blob/master/README.rst), cause that means users will see this:

image

instead of the current version number, when loading the sdk-nrf github page: https://github.com/nrfconnect/sdk-nrf

is that what we want ?
> Plus "Update documentation version for the different doc sets (in the doc/*/conf.py files)" - we also have versions for mcuboot and nrfxlib, but those weren't updated for 1.3.2.

as mcuboot and nrfxlib doesn't contain a `VERSION` file today, we can't do the same for those currently :disappointed:

**Comment by ru-fu** [comment]
> One question, why is this needed:
> 
> https://github.com/nrfconnect/sdk-nrf/blob/343b938e0ae7e3ad4b77b05be9481901fae51d3e/doc/nrf/shortcuts.txt#L3
> 
> when we have `conf.py` ?

We need this to be compatible with master. The version on master is "1.3.99", but we cannot use that version number where we are talking about a release (for example, we cannot check out the west revision for 1.3.99, but we need to use the latest release).
So I guess we can't fill this one automatically.

> One observation I did is this line:
> 
> ```
> for development in |release| for prototyping
> ```
> 
> in [README.rst](https://github.com/nrfconnect/sdk-nrf/blob/master/README.rst), cause that means users will see this:
> 
> image
> 
> instead of the current version number, when loading the sdk-nrf github page: https://github.com/nrfconnect/sdk-nrf
> 
> is that what we want ?

Please agree with @joerchan - we can't do automatic replacements without using shortcuts. ;)

> as mcuboot and nrfxlib doesn't contain a `VERSION` file today, we can't do the same for those currently 😞

:( Those are the ones that are most frequently forgotten.

**Comment by joerchan** [comment]
> Please agree with @joerchan - we can't do automatic replacements without using shortcuts. ;)

Don't have to agree with me. I was just making suggestions :)

It does seem that it is wrong, as @tejlmand points out. It does not say the release version but instead it says `|release|` and a link to  nowhere.

**Comment by tejlmand** [comment]
> We need this to be compatible with master. The version on master is "1.3.99", but we cannot use that version number where we are talking about a release (for example, we cannot check out the west revision for 1.3.99, but we need to use the latest release).

understood, but isn't it dangerous to also call it `release`.
That will hide `release` from `conf.py` won't it ?
Should it maybe have been called: `latest_release`, instead ?

**Comment by tejlmand** [comment]
@joerchan and @ru-fu should I add a commit in this PR that reverts `|release|` back to `v1.3.2` in https://github.com/nrfconnect/sdk-nrf/blob/master/README.rst ?
or just wait until tagging `rc` to be updated there ?

**Comment by ru-fu** [comment]
> 
> 
> > We need this to be compatible with master. The version on master is "1.3.99", but we cannot use that version number where we are talking about a release (for example, we cannot check out the west revision for 1.3.99, but we need to use the latest release).
> 
> understood, but isn't it dangerous to also call it `release`.
> That will hide `release` from `conf.py` won't it ?
> Should it maybe have been called: `latest_release`, instead ?

Up to now, version and release in conf.py have always been the same. I don't know in which cases these will differ now and which we'll then want to use.

**Comment by tejlmand** [comment]
> Up to now, version and release in conf.py have always been the same. I don't know in which cases these will differ now and which we'll then want to use.

Well, the comment says:
```
# The short X.Y version.
version = '1.3.99'
# The full version, including alpha/beta/rc tags.
release = '1.3.99'
```
So I just did what the comment stated, so if `VERSION` contains:
```
1.4.0-rc1
```
then `version==1.4.0` and `release==1.4.0-rc1`, as the comment states.

@ru-fu let me know if that should be changed before merging this.

**Comment by ru-fu** [comment]
> then `version==1.4.0` and `release==1.4.0-rc1`, as the comment states.

If "version" is what is then displayed in the top-left corner, this is how we've been doing it up to now, so that is fine.
And in this case, we'll never use the "release" from conf.py, and can safely override it with the release that actually makes sense for the docs.

**Comment by tejlmand** [comment]
> > then `version==1.4.0` and `release==1.4.0-rc1`, as the comment states.
> 
> If "version" is what is then displayed in the top-left corner, this is how we've been doing it up to now, so that is fine.

Just did a local build where `-rc1` was appended for additional verification, and it showed up correctly 😄

**Pull request merged**